### PR TITLE
[Feat] Populating nested mutations with other relationships for @create

### PR DESCRIPTION
**Pull request by liepaja**
- [x] Added or updated tests
- [x] Added Docs for all relevant versions

**Related Issue**

#396 #332

**Changes**

Adding other relationship capabilities to nested mutation.

| **T** indicates presence of a test

Existing nested mutation relationships from #207:
`@create`:
- **BelongsTo**
- [x] no arg | **T** moved to connect
- **HasMany**
- [x] create | **T**

Nested mutation relationships from current #514:
`@create`:
------------------------------- One to one/many
- **BelongsTo**
- [x] create | **T**
- [x] connect | **T**
- **HasOne**
- [x] create | **T**
------------------------------- Many to many
- **BelongsToMany**
- [x] create | **T**
- [x] connect | **T**
------------------------------- Morph one to one/many
- **MorphOne**
- [x] create | **T**
- **MorphMany**
- [x] create | **T**
- **MorphTo**
- [x] no args | **T**

**Breaking changes**

Instead of passing existing belongsTo ID as variable, it should be wrapped in a `connect` argument.

```graphql
type User {
  ....
}

type Post {
  id: ID!
  title: String!
  author: User! @belongsTo
}

input CreatePostRelation {
  create: [CreatePostInput!]
}
```
From:
```graphql
input CreatePostInput {
  title: String!
  author: ID
}
```
To:
```graphql
input CreatePostInput {
  title: String!
  author: {
   connect: ID || connect: [ID] / based on many to many or one to many
  }
}
```

### CODE REVIEW

**Comment by spawnia** [comment]
@liepaja Very impressed with your methodical approach to this.

Would love for nested mutations to be more powerful!

I do disagree with the breaking change about the `BelongsTo` relationship. The name of the underlying column should not matter to the end user of the schema.

In the example you made, it seems reasonable as the foreign key has a great speaking name. It falls apart as soon as the relation was renamed or the application was possibly built on a legacy database schema.

If the column name was `user_id` but the relation is named `author`, it is much clearer to name the input after the relation. The column might be named even worse, something like `post_author_uuid_foreign`. Would you still want to use that in the schema?

Eloquent does a good job of hiding such gory implementation details, so we should leverage that!

File: src/Execution/MutationExecutor.php [hunk 1]
```diff
@@ -2,25 +2,39 @@
 
 namespace Nuwave\Lighthouse\Execution;
 
+use Illuminate\Database\Eloquent\Relations\BelongsToMany;
```
Comment (spawnia) [inline src/Execution/MutationExecutor.php]: Please pay attention to the import order. `php-cs-fixer` can do that automatically.

File: src/Execution/MutationExecutor.php [hunk 2]
```diff
@@ -2,25 +2,39 @@
...
 use Illuminate\Database\Eloquent\Model;
 use Illuminate\Database\Eloquent\Relations\HasMany;
 use Illuminate\Database\Eloquent\Relations\BelongsTo;
 use Illuminate\Database\Eloquent\ModelNotFoundException;
 
+
 class MutationExecutor
 {
+
+
```
Comment (spawnia) [inline src/Execution/MutationExecutor.php]: No extranous blank lines

File: src/Execution/MutationExecutor.php [hunk 3]
```diff
@@ -48,7 +80,7 @@ public static function executeCreate(Model $model, Collection $args, HasMany $pa
      *
      * @return Model
      */
-    protected static function saveModelWithBelongsTo(Model $model, Collection $remaining, HasMany $parentRelation = null): Model
+    protected static function saveModelWithBelongsTo(Model $model, Collection $remaining, Relation $parentRelation = null): Model
```
Comment (spawnia) [inline src/Execution/MutationExecutor.php]: Make sure to update the docblocks as well

File: src/Execution/MutationExecutor.php [hunk 4]
```diff
@@ -77,14 +117,23 @@ protected static function saveModelWithBelongsTo(Model $model, Collection $remai
     protected static function handleHasManyCreate(Collection $multiValues, HasMany $relation)
     {
         $multiValues->each(function ($singleValues) use ($relation) {
-            self::executeCreate(
-                $relation->getModel()->newInstance(),
-                collect($singleValues),
-                $relation
-            );
+            self::executeCreate($relation->getModel()->newInstance(), collect($singleValues), $relation);
```
Comment (spawnia) [inline src/Execution/MutationExecutor.php]: Such complicated method calls should be split into multiple lines.

This helps readability and the ability to debug issues. For example, if `$relation->getModel()->newInstance()` fails, you would get an exact line number pointing to that.

File: src/Execution/MutationExecutor.php [hunk 5]
```diff
@@ -107,22 +156,18 @@ public static function executeUpdate(Model $model, Collection $args, HasMany $pa
 
         $model = self::saveModelWithBelongsTo($model, $remaining, $parentRelation);
 
-        $hasMany->each(function ($nestedOperations, string $relationName) use ($model) {
+        $hasMany->each(function ($nestedOperations, $relationName) use ($model) {
```
Comment (spawnia) [inline src/Execution/MutationExecutor.php]: Can this ever be anything else then a string?

File: src/Execution/MutationExecutor.php [hunk 6]
```diff
@@ -156,9 +201,35 @@ public static function executeUpdate(Model $model, Collection $args, HasMany $pa
...
+
+    /**
+     * @param Model $model
+     * @param Collection $args
+     *
+     * @return Collection
+     */
+    protected static function extractMorphToArgs(Model $model, Collection $args): Collection
+    {
+        return $args->partition(function ($value, $key) use ($model) {
```
Comment (spawnia) [inline src/Execution/MutationExecutor.php]: Consider extracting the partitioning into its own method and pass the type of relation as an argument.

File: tests/Utils/Models/Hour.php [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+
+namespace Tests\Utils\Models;
+
+use Illuminate\Database\Eloquent\Model;
+use Illuminate\Database\Eloquent\Relations\MorphTo;
+
+class Hour extends Model
+{
+    protected $fillable = ['from', 'to', 'weekday'];
+
```
Comment (spawnia) [inline tests/Utils/Models/Hour.php]: no extranous blank lines

File: src/Execution/MutationExecutor.php (hunk 5 above)
Comment (spawnia) [reply src/Execution/MutationExecutor.php]: Why was this type hint removed?

File: src/Execution/MutationExecutor.php [hunk 8]
```diff
@@ -107,26 +187,22 @@ public static function executeUpdate(Model $model, Collection $args, HasMany $pa
...
         $model = self::saveModelWithBelongsTo($model, $remaining, $parentRelation);
 
-        $hasMany->each(function ($nestedOperations, string $relationName) use ($model) {
+        $hasMany->each(function ($nestedOperations, $relationName) use ($model) {
             /** @var HasMany $relation */
             $relation = $model->{$relationName}();
 
-            collect($nestedOperations)->each(function ($values, string $operationKey) use ($relation) {
-                if ($operationKey === 'create') {
+            collect($nestedOperations)->each(function ($values, $operationKey) use ($relation) {
```
Comment (spawnia) [inline src/Execution/MutationExecutor.php]: Why did you remove the `string` type hint?

Comment (liepaja) [reply src/Execution/MutationExecutor.php]: Messed up branches, this is old commit

**Comment by liepaja** [comment]
Any suggestions about MorphTo?

**Comment by liepaja** [comment]
Currently when having an update mutation, apollo attaches __typename. What do you think @spawnia , should it be handled, or users must put this manually into input object?

**Comment by spawnia** [comment]
@liepaja i am not quite sure what you are saying. Can you eloborate with a concrete example?

**Comment by spawnia** [comment]
Apollo adds the `__typename` field to queries automatically, not the Inputs. If you directly use the results of a query as the arguments for a mutation, you might run into the problem you are experiencing.

While the GraphQL spec does reserve `__typename` for query fields, it is a perfectly valid name for an actual input argument, so we should not do any special treatment on the server side.

We ran into this issue in our work project, our solution was to omit the `__typename` fields from a query result before passing them into a mutation as inputs. https://github.com/jonschlinkert/omit-deep works quite well for us.

**Comment by liepaja** [comment]
Yeah using json from query was my case. Didn't track it prior. So yeah you are right

**Comment by liepaja** [comment]
@spawnia StyleCI is nice, but how do you automate it?

**Comment by spawnia** [comment]
@liepaja you can click on the StyleCI check, download the diff file and `git apply` it. If you use PHPStorm, just drag-and-drop it into the editor.

**Comment by spawnia** [comment]
Nice job @liepaja

How about `@update`, have something planned for that as well?

**Comment by liepaja** [comment]
@spawnia i decided to split pull requests, as it would get to unstructured. After this will be merged will open one for @update

**Comment by chrissm79** [comment]
@liepaja Superb work on this!!! Did you want to merge in master and fix the conflict? I could do it if you prefer but it's your PR so it's up to you. Thanks!!

**Comment by liepaja** [comment]
@chrissm79 resolved
